cli/src/cmd/run.rs

```diff
@@ -89,168 +88,127 @@ impl Cmd for RunArgs {
-        let bytecode = bytecode.expect("No bytecode").object.into_bytes().unwrap();
+        let abi = abi.expect("no ABI for contract");
+        let bytecode = bytecode.expect("no bytecode for contract").object.into_bytes().unwrap();
         let needs_setup = abi.functions().any(|func| func.name == "setUp");
 
-        let mut cfg = crate::utils::sputnik_cfg(&evm_version);
-        cfg.create_contract_limit = None;
-        let vicinity = evm_opts.vicinity()?;
-        let backend = evm_opts.backend(&vicinity)?;
-
-        // need to match on the backend type
-        let result = match backend {
-            BackendKind::Simple(ref backend) => {
-                let runner = ContractRunner::new(
-                    &evm_opts,
-                    &cfg,
-                    backend,
-                    &abi,
-                    bytecode,
-                    Some(evm_opts.sender),
-                    None,
-                    &predeploy_libraries,
-                );
-                runner.run_test(&func, needs_setup, Some(&known_contracts))?
-            }
-            BackendKind::Shared(ref backend) => {
-                let runner = ContractRunner::new(
-                    &evm_opts,
-                    &cfg,
-                    backend,
-                    &abi,
-                    bytecode,
-                    Some(evm_opts.sender),
-                    None,
-                    &predeploy_libraries,
-                );
-                runner.run_test(&func, needs_setup, Some(&known_contracts))?
-            }
+        let mut builder = ExecutorBuilder::new()
+            .with_cheatcodes(evm_opts.ffi)
+            .with_config(evm_opts.evm_env())
+            .with_spec(crate::utils::evm_spec(&config.evm_version));
+        if let Some(ref url) = self.evm_opts.fork_url {
+            let fork = Fork { url: url.clone(), pin_block: self.evm_opts.fork_block_number };
+            builder = builder.with_fork(fork);
+        }
+        if verbosity >= 3 {
+            builder = builder.with_tracing();
+        }
+        if evm_opts.debug {
+            builder = builder.with_tracing().with_debugger();
+        }
+
+        let mut result = {
+            let mut runner =
+                Runner::new(builder.build(), evm_opts.initial_balance, evm_opts.sender);
+            let (address, mut result) =
+                runner.setup(&predeploy_libraries, bytecode, needs_setup)?;
+
+            let RunResult {
+                success,
+                gas_used,
+                logs,
+                traces,
+                debug: run_debug,
+                labeled_addresses,
+                ..
+            } = runner
+                .run(address, encode_args(&IntoFunction::into(self.sig), &self.args)?.into())?;
+
+            result.success &= success;
+
+            result.gas_used = gas_used;
+            result.logs.extend(logs);
+            result.traces.extend(traces);
+            result.debug = run_debug;
```

Comment:
We can extend the debugger to also debug constructors in contracts and setup functions by changing this line. I'm thinking we probably want to add some sort of metadata to the debug arenas that denote what step they are in (similar to traces). We can then add some selector UI in the debugger to skip parts you don't care about.